src/compiler/program.ts

```diff
@@ -3805,7 +3890,12 @@ export function createProgram(rootNamesOrOptions: readonly string[] | CreateProg
     }
 
     function addFileIncludeReason(file: SourceFile | undefined, reason: FileIncludeReason) {
-        if (file) fileReasons.add(file.path, reason);
+        if (file) {
+            const existing = fileReasons.get(file.path);
+            if (!some(existing, r => fileIncludeReasonIsEqual(r, reason))) {
+                fileReasons.add(file.path, reason);
+            }
+        }

```

Comment:
I didn't get to it in #58352, but is it worth not doing this here and instead post-process the reasons once at the end, to prevent this from blowing up time-wise? I'm not sure how long these lists can end up getting.

Reply:
moving this as post processing needs to handle it for file processing diagnostic deduplication as well, making it more complicated than it already is

Reply:
I think a better way to do this is to see if file references have already been processed. I changed that.